File: 0496-next-greater-element-i/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Stack<Integer>stk = new Stack<>();
        int n = nums1.length, m =  nums2.length;
        int []ans = new int[n];
        stk.push(-1);
        for(int j = 0;j < n;j++){
            for(int i = m - 1;i >= 0;i--){
                if(stk.size() > 0){
                   int top = stk.peek();
                if(nums2[i] > nums1[j]){
                  stk.pop();
                  stk.push(nums2[i]);
                }
                else if(nums2[i] == nums1[j]) break;

                }
            }
            if(stk.size() > 0){
            ans[j] = stk.pop();
            stk.push(-1);
            }
            
        }
        return ans;
    }
}

File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        int n = nums2.size(), m = nums1.size();
        unordered_map<int, int>nge;
        stack<int>stk;
        nge[nums2[n - 1]] = -1;
        stk.push(nums2[n - 1]);
        for(int i = n - 2;i >= 0;i--){
            if(nums2[i] > stk.top()){
              while(!stk.empty() && nums2[i] > stk.top()){
                stk.pop();
              }
            }
            if(stk.empty()){
             nge[nums2[i]] = -1;   
             stk.push(nums2[i]);
              continue;    
            }
             nge[nums2[i]] = stk.top();
            stk.push(nums2[i]);
        }
        vector<int>ans(m);
        for(int i = 0;i < m;i++){
            ans[i] = nge[nums1[i]];
        }
        return ans;
    }
};